pennylane_forest/converter.py

```diff
@@ -0,0 +1,626 @@
+    sum[: A.shape[0], : A.shape[1]] = A
+    sum[A.shape[0] :, A.shape[1] :] = B
+
+    return sum
+
+
+def _controlled_matrix(op):
+    """Return the matrix associated with the controlled operation.
+    
+    Args:
+        op (np.array): Array representing the operation 
+                       that should be controlled
+    
+    Returns:
+        np.array: Array representing the controlled operations. If the input
+                  array has shape (N, N) the output shape is (2*N, 2*N).
+    """
+    return _direct_sum(np.eye(op.shape[0], dtype=op.dtype), op)
+
+
+def _resolve_gate(gate):
+    """Resolve the given pyquil Gate as far as possible.
+
+    For example, the gate `CONTROLLED CONTROLLED X` will be resolved to `CCNOT`. 
+    The gate `CONTROLLED CONTROLLED RX(0.3)` will be resolved to `CONTROLLED CRX(0.3)`.
+    
+    Args:
+        gate (pyquil.quil.Gate): The gate that should be resolved
+    
+    Returns:
+        pyquil.quil.Gate: The maximally resolved gate
+    """
+    for i, modifier in enumerate(gate.modifiers):
+        if modifier == "CONTROLLED":
+            if gate.name in _control_map:
+                stripped_gate = g.Gate(_control_map[gate.name], gate.params, gate.qubits)
+                stripped_gate.modifiers = gate.modifiers.copy()
+                del stripped_gate.modifiers[i]
+
+                return _resolve_gate(stripped_gate)
+            else:
+                break
+
+    return gate
+
+
+def _resolve_params(params, variable_map):
+    """Resolve a parameter list with a given variable map.
+    
+    Args:
+        params (List[Union[pyquil.quilatom.MemoryReference, object]]): The parameter list
+        variable_map (Dict[str, object]): The map that assigns values to variable names
+    
+    Returns:
+        List[object]: The resolved parameters. This list does not contain MemoryReferences anymore.
+    """
+    resolved_params = []
+
+    for param in params:
+        if isinstance(param, pyquil.quilatom.MemoryReference):
+            resolved_params.append(variable_map[param.name])
+        else:
+            resolved_params.append(param)
+
+    return resolved_params
+
+
+def _normalize_variable_map(variable_map):
+    """Normalize the given variable map.
+
+    Variable maps can have keys that are either strings or
+    pyquil.quil.MemoryReference instances. This methods replaces
+    all MemoryReference instances with their name.
+    
+    Args:
+        variable_map (Dict[Union[str, pyquil.quil.MemoryReference], object]): The initial variable map.
+    
+    Returns:
+        Dict[str, object]: Variable map with all MemoryReference instances replaced by their name
+    """
+    new_keys = list(variable_map.keys())
+    values = list(variable_map.values())
+
+    for i in range(len(new_keys)):
+        if isinstance(new_keys[i], pyquil.quil.MemoryReference):
+            new_keys[i] = new_keys[i].name
+
+    return dict(zip(new_keys, values))
```

Comment:
You could potentially remove the `(key,value)` pair here and then add the `(key.name, value)` here. This way there would be no need to reconstruct the dictionary.